### ISSUE TITLE
Adding APIs to expose valid data size information at a store level

### ISSUE THREAD

**Issue by xunyin8** [linkedin/ambry#546]
We currently have no API in ambry to provide space consumption statistics. This is a problem because statistics like valid data size per log segment is needed by compaction. Current valid data size per serviceId and containerId is also needed by quota management and containerization of ambry in the future.

Additions to BlobStore
1. BlobStoreStats: this class is responsible for the collection, maintenance and service of valid data size stats. The control logic for deciding whether a blob is valid or not also sits in here.
   - Collection: Scan the index and create/update appropriate data structures to store relevant stats
   - Maintenance: Keep stats collected by a previous scan up-to-date by processing new writes and make appropriate changes.
   - Service: Serving stats requests by using previously collected stats or trigger a new scan if requested stats is not readily available.

2. ScanResults: this class is a data class that holds all data structure needed by BlobStoreStats in order to serve stats related requests. This class contains various helper functions to update its fields and access the data stored in this class. This data class is populated during scanning and accessed when serving stats requests.

Bucket concept (time slot): this is used to consolidate the data changes in the store over some time span into a single data point. For example, if a 10MB blob is expired at 1:15PM, a 20MB blob is expired at 1:30PM and a 35MB blob is expired at 1:50PM then a Bucket could be constructed with a value of -65MB. The Bucket will also have a time stamp, in our example the time stamp could be 2:00PM indicating that by the end of 2:00PM a total of 65MB worth of blobs have expired. Data structures for these buckets are inside ScanResults.

3. StatsEvent: this class is used for the purpose of handling new puts and deletes during a scan. StatsEvent captures/buffers all needed information of puts and deletes that happen while a scan is in progress. Once the scan is finished these buffered StatsEvent will then be processed and added to the data structures accordingly.

[Gist](https://gist.github.com/xunyin8/86779d18d616cc2d665ee58914a48bcb) for the APIs.

**Comment by vgkholla** [comment]
@xunyin8 Could this be added #530 instead of creating a new issue?

**Comment by xunyin8** [comment]
@vgkholla I think this is different enough to #530 that it will be a complete rewrite. Do you still want me to do that?
